socs/testing/hwp_emulator.py

```diff
@@ -0,0 +1,223 @@
+"""
+HWP Emulation module
+"""
+import logging
+import threading
+import time
+from dataclasses import dataclass
+
+import pytest
+
+from socs.agents.hwp_pid.drivers.pid_controller import PID
+from socs.testing import device_emulator
+
+
+def hex_str_to_dec(hex_value, decimal=3):
+    """Converts a hex string to a decimal float"""
+    return float(int(hex_value, 16)) / 10**decimal
+
+
+@dataclass
+class PMXState:
+    """State of the PMX Emulator"""
+    output: bool = False
+    current: float = 0
+    current_limit: float = 10.0
+    voltage_limit: float = 10.0
+    voltage: float = 0
+    source: str = "volt"
+
+
+@dataclass
+class PIDState:
+    """State of the PID Emulator"""
+    direction: str = "forward"
+    freq_setpoint: float = 0.0
+
+
+@dataclass
+class HWPState:
+    """State of the HWP Emulator"""
+    cur_freq: float = 0.0
+    pmx: PMXState = PMXState()
+    pid: PIDState = PIDState()
+    lock = threading.Lock()
+
+
+def _create_logger(name, log_level=logging.INFO):
+    logger = logging.getLogger(name)
+    logger.setLevel(log_level)
+    if len(logger.handlers) == 0:
+        formatter = logging.Formatter("%(name)s: %(message)s")
+        handler = logging.StreamHandler()
+        handler.setFormatter(formatter)
+        logger.addHandler(handler)
+    return logger
+
+
+def lerp(start, end, t):
+    return (1 - t) * start + t * end
+
+
+class HWPEmulator:
+    def __init__(self, pid_port=None, pmx_port=None, log_level=logging.INFO):
+        self.pid_port = pid_port
+        self.pmx_port = pmx_port
+
+        self.state = HWPState()
+
+        self.pid_device = device_emulator.DeviceEmulator([])
+        self.pid_device.get_response = self.process_pid_msg
+        self.pid_device.logger = _create_logger("PID", log_level=log_level)
+
+        self.pmx_device = device_emulator.DeviceEmulator([])
+        self.pmx_device.get_response = self.process_pmx_msg
+        self.pmx_device.logger = _create_logger("PMX", log_level=log_level)
+
+        self.update_thread = threading.Thread(target=self.update_loop)
+        self.run_update = False
+
+        self.logger = _create_logger("HWP", log_level=log_level)
+
+    def start(self):
+        """Start up TCP Sockets and update loop"""
+        if self.pid_port is not None:
+            self.pid_device.create_tcp_relay(self.pid_port)
+        if self.pmx_port is not None:
+            self.pmx_device.create_tcp_relay(self.pmx_port)
+
+        self.update_thread.start()
+
+    def shutdown(self):
+        """Shutdown TCP Sockets and update loop"""
+        self.run_update = False
+        self.pid_device.shutdown()
+        self.pmx_device.shutdown()
+        self.update_thread.join()
```

Comment:
This seems like it might shutdown the devices before the `update_loop` thread terminates. Wouldn't you want to `.join()`, ensuring any communication executed by the update thread finishes, before `.shutdown()`?

Reply:
The update thread is just for state updates, so we don't really care about consistency or final execution on shutdown.

Reply:
Ah, yup, that makes sense. Thought it might have been a possible source of some communication after shutdown, which would result in that "bad file descriptor" we've seen.